builtin/providers/openstack/resource_openstack_compute_secgroup_v2.go

```diff
@@ -0,0 +1,294 @@
+package openstack
+
+import (
+	"bytes"
+	"fmt"
+	"log"
+
+	"github.com/hashicorp/terraform/helper/hashcode"
+	"github.com/hashicorp/terraform/helper/schema"
+	"github.com/rackspace/gophercloud"
+	"github.com/rackspace/gophercloud/openstack/compute/v2/extensions/secgroups"
+)
+
+func resourceComputeSecGroupV2() *schema.Resource {
+	return &schema.Resource{
+		Create: resourceComputeSecGroupV2Create,
+		Read:   resourceComputeSecGroupV2Read,
+		Update: resourceComputeSecGroupV2Update,
+		Delete: resourceComputeSecGroupV2Delete,
+
+		Schema: map[string]*schema.Schema{
+			"region": &schema.Schema{
+				Type:        schema.TypeString,
+				Required:    true,
+				ForceNew:    true,
+				DefaultFunc: envDefaultFunc("OS_REGION_NAME"),
+			},
+			"name": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"description": &schema.Schema{
+				Type:     schema.TypeString,
+				Required: true,
+				ForceNew: false,
+			},
+			"rule": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"id": &schema.Schema{
+							Type:     schema.TypeString,
+							Computed: true,
+						},
+						"from_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+							ForceNew: false,
+						},
+						"to_port": &schema.Schema{
+							Type:     schema.TypeInt,
+							Required: true,
+							ForceNew: false,
+						},
+						"ip_protocol": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+							ForceNew: false,
+						},
+						"cidr": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: false,
+						},
+						"from_group_id": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							ForceNew: false,
+						},
+						"self": &schema.Schema{
+							Type:     schema.TypeBool,
+							Optional: true,
+							Default:  false,
+							ForceNew: false,
+						},
+					},
+				},
+			},
```

Comment:
We're learning from feedback on the AWS provider that it may be better in the long run for us to break out security group rules into their own top-level resource. This allows:

 * Richer dependency graphs to be expressed for more complicated update scenarios
 * Rules to be added to security groups that may not be fully managed by terraform

Since it looks like the backing APIs actually treat Rules as a first class object already, it may actually make the provider code more straightforward. What do you think?